Added new class for  Ansible Runner Callbacks

amolgautam25 opened the pull request:
<!--- changelog-entry
# Fill in 'msg' below to have an entry automatically added to the next release changelog.
# Leaving 'msg' blank will not generate a changelog entry for this PR.
# Please ensure this is a simple (and readable) one-line string.
---
msg: ""
-->

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 16.0.1.dev3184+g2f0e80f113
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```


Things to do: 

- [x] Figure out a way to update the ```instance``` from db in the new class and propagate it back to the BaseTask
- [x] Fix unit tests

AlanCoding commented on awx/main/tasks/jobs.py:
```diff
@@ -76,6 +76,177 @@
+            if cancel_wait > 5:
+                logger.warn('Request to cancel {} took {} seconds to complete.'.format(self.instance.log_format, cancel_wait))
+            return True
+        return False
```
With the other PR, this will be handled in a different class, but I think we will be free of conflicts if we just leave this method unchanged in this diff. We can work on an elegant combination of the 2 in another unit of work.

AlanCoding commented on awx/main/tasks/jobs.py:
```diff
@@ -76,6 +76,177 @@
 logger = logging.getLogger('awx.main.tasks.jobs')
 
 
+class RunnerCallback:
```
I don't see this used anywhere?

You have references like `self.event_data_key` in here. In that, `self` is supposed to be a subclass of `BaseTask` and that's not true when you split it into a different class. So you can process these in an `__init__` method, and if it makes sense to you, save it right to `self` again. This clarifies the data flow between the task class and the callback class.

And obviously, you'll have to initialize it somewhere and replace the references to the methods.

amolgautam25 replied:
@AlanCoding , thank you.
I have tried to make the required changes.

AlanCoding commented on awx/main/tasks/jobs.py:
```diff
@@ -82,6 +82,190 @@
+        self.host_map = BaseTaskObject.host_map
+        self.guid = BaseTaskObject.guid
+        self.recent_event_timings = BaseTaskObject.recent_event_timings
+        self.instance = BaseTaskObject.instance
```
style-wise, `BaseTaskObject` looks like it should be a class, I'd prefer a style like `task_object` so it's more obviously an instantiated object.

amolgautam25 commented on awx/main/tasks/callback.py:
```diff
@@ -0,0 +1,188 @@
+        # ansible-inventory and the awx.main.commands.inventory_import
+        # logger
+        if isinstance(self, RunInventoryUpdate):
+            self.end_line = event_data['end_line']
```
Right now there are variables such as ```end_line``` which belong to class ```BaseTask```. I am not sure if these variables are supposed to be moved to ```RunnerCallback``` class or left as is.

amolgautam25 replied:
https://github.com/ansible/awx/pull/11571/files#r793020460

amolgautam25 commented on awx/main/tasks/receptor.py:
```diff
@@ -405,15 +405,20 @@ def transmit(self, _socket):
             _socket.shutdown(socket.SHUT_WR)
 
+    def attach_runner_callback(self):
+        from awx.main.tasks.jobs import RunnerCallback
```
To avoid circular imports.

amolgautam25 replied:
This is no longer relevant

amolgautam25 commented on awx/main/tasks/jobs.py:
```diff
@@ -444,164 +442,8 @@ def final_run_hook(self, instance, status, private_data_dir, fact_modification_t
-                with disable_activity_stream():
-                    self.instance = self.update_model(self.instance.pk, result_traceback=result_traceback)
+    def attach_callback(self):
+        self.runner_callback = RunnerCallback()
```
The two classes ```BaseTak``` and ```RunnerCallback``` are coupled here.

amolgautam25 replied:
If we look at this coupling we can easily do something like this :

--- change a value inside of a callback instance
example : self.runner_callback.some_variable = new_value

What I am stuck at is the following:
1) if something  changes inside the runner_callback ( which is kind of an inside class, how would that change be propagated to  outside class, in this case ```BaseTask``` . 
I can go ahead and implement something like this: https://stackoverflow.com/questions/2024566/how-to-access-outer-class-from-an-inner-class , but it seems to overcomplicate things. 

2) Where do I invoke this function ? There has to be a place where a runner_callback is attached to the BaseTask.

amolgautam25 replied:
I have removed the ```attach callback()``` function

AlanCoding commented on awx/main/tasks/jobs.py:
```diff
@@ -107,12 +105,15 @@ class BaseTask(object):
     abstract = True
```
Right here I would prefer to do

```python
    callback_class = RunnerCallback
```

Then in RunProjectUpdate, put in the respective callback class

amolgautam25 replied:
I have made the required change.

AlanCoding commented on awx/main/tasks/jobs.py:
```diff
@@ -107,12 +105,15 @@ class BaseTask(object):
     abstract = True
 
     def __init__(self):
+        self.runner_callback = RunnerCallback()
```
Corresponding to prior comment, change this to 

```python
        self.runner_callback = self.callback_class()
```

AlanCoding commented on awx/main/tasks/jobs.py:
```diff
@@ -107,12 +105,15 @@ class BaseTask(object):
+        self.runner_callback.guid = GuidMiddleware.get_guid()
+        self.runner_callabck.job_created = None
+        self.runner_callback.recent_event_timings = deque(maxlen=settings.MAX_WEBSOCKET_EVENT_RATE)
+        self.runner_callback.instance.id = self.instance.id
```
could easily add

```python
     self.runner_callback.event_data_key = self.event_data_key
```

amolgautam25 replied:
I have added that.

AlanCoding commented on awx/main/tasks/jobs.py:
```diff
@@ -1203,18 +1048,11 @@ class RunProjectUpdate(BaseTask):
         super(RunProjectUpdate, self).__init__(*args, **kwargs)
-        self.playbook_new_revision = None
+        self.runner_callback = RunnerCallbackForProjectUpdate()
+        self.runner_callback.playbook_new_revision = None
```
Setting `playbook_new_revision` to None would be better done in `RunnerCallbackForProjectUpdate.__init__`

amolgautam25 replied:
I have made the required changes.

amolgautam25 committed:
Decoupled callback functions from BaseTask Class

--- Removed all callback functions from 'jobs.py' and put them in a new file '/awx/main/tasks/callback.py'
--- Modified Unit tests unit moved
--- Moved 'update_model' from jobs.py to /awx/main/utils/update_model.py

amolgautam25 commented on awx/main/tasks/callback.py:
```diff
@@ -0,0 +1,203 @@
+        # ansible-inventory and the awx.main.commands.inventory_import
+        # logger
+        if isinstance(self, RunInventoryUpdate):
+            # Will removing this 'if'statement from here have any performance impact ?
```
@AlanCoding In this new file this line ``` if isinstance(self, RunInventoryUpdate):``` would always return false. 
Would it be wise to just remove this line? 
Any thoughts / suggestion on this ?

AlanCoding replied:
Just like you introduced a class `RunnerCallbackForProjectUpdate`, you can cover this elegantly by introducing `RunnerCallbackForInventoryUpdate`, then move the logic there, and then delete it here.

You could keep it here and then change references in `jobs.py` to `self.runner_callback.end_line`, if that makes it easier to get to a state of functionality.

amolgautam25 commented on awx/main/tasks/callback.py:
```diff
@@ -0,0 +1,203 @@
+            if event_data.get('event') == 'playbook_on_stats':
+                event_data['host_map'] = self.host_map
+
+        if isinstance(self, RunProjectUpdate):
```
@AlanCoding following up on https://github.com/ansible/awx/pull/11571/files#r795575348 , i can see a similar problem here. 
Can I do something like this : 
```python
 if isinstance(self,RunnercallbackForProjectUpdate)
```

This seems logical to me. I think it would work here.

AlanCoding replied:
Yes, I agree.

amolgautam25 commented on awx/main/tasks/callback.py:
```diff
@@ -0,0 +1,203 @@
+        Ansible runner callback to tell the job when/if it is canceled
+        """
+        unified_job_id = self.instance.pk
+        self.instance = self.update_model(unified_job_id)
```
```self.instance and self.update_model``` are both not accessible here. 
I have a bit of idea how we can change the model from the database. ( ```refresh_from_db()``` ) , but i am not sure how we  can use that here  to update the ```instance```.

On top of that ```instance``` belongs to the ```BaseTask``` class. I have not yet removed the reference of that from callback class. 

@AlanCoding  any thoughts on this one ?

AlanCoding replied:
Yes, the current cancel_callback is fundamentally dependent on having the instance available and accessing the database. Until that is refactored, `self.instance` will need to be kept on the callback class. To do that, you have to assign it in `__init__` which shouldn't be a problem since it receives the `BaseTask` derived object to begin with.

AlanCoding commented on awx/main/tasks/callback.py:
```diff
@@ -0,0 +1,203 @@
+        self.runner_callback.host_map = {}
+        self.runner_callback.guid = GuidMiddleware.get_guid()
+        self.runner_callback.job_created = None
+        self.runner_callback.recent_event_timings = deque(maxlen=settings.MAX_WEBSOCKET_EVENT_RATE)
```
Pattern-wise, `self.runner_callback.parent_workflow_job_id = None` --> `self.parent_workflow_job_id = None`

This is possibly the most important thing to clear up so that you can get running with testing this.

amolgautam25 replied:
I have made the changes.

AlanCoding commented on awx/main/tasks/jobs.py:
```diff
@@ -105,14 +103,12 @@ class BaseTask(object):
-        self.guid = GuidMiddleware.get_guid()
-        self.job_created = None
-        self.recent_event_timings = deque(maxlen=settings.MAX_WEBSOCKET_EVENT_RATE)
+        self.runner_callback.event_data_key = self.event_data_key
```
As I search this file for `event_data_key`, I don't see any reason the task class should own it. I feel that it should be moved entirely to the callback class. So, I picture:

```python
class RunnerCallbackForProjectUpdate(RunnerCallback):
    event_data_key = 'project_update_id'
```

That means that you would have to introduce a new class for every job type that has an event_data_key defined. To me, that feels like the right solution. That will clean up a lot of the work in `__init__` methods here.

AlanCoding commented on awx/main/tasks/jobs.py:
```diff
@@ -723,16 +561,16 @@ def run(self, pk, **kwargs):
                     del params[v]
 
-            self.dispatcher = CallbackQueueDispatcher()
+            self.runner_callback.dispatcher = CallbackQueueDispatcher()
```
I can't see any reason for `CallbackQueueDispatcher` to be in this file. I think it should be moved to the callback module. It should work in the `__init__` method.

amolgautam25 replied:
I have made the required changes.

AlanCoding commented on awx/main/tasks/jobs.py:
```diff
@@ -802,6 +640,9 @@ class RunJob(BaseTask):
     event_data_key = 'job_id'
 
+    def __init__(self):
+        self.runner_callback.event_data_key = self.event_data_key
```
I expect this will break inventory update runs, because you didn't defer to the `super` method. The proper way to override `__init__` in a subclass is that you call the super's method (like `super(self, RunJob).__init__(*args, **kwargs)`).

But if you take my advice from a prior comment, you shouldn't need this method at all.

AlanCoding commented on awx/main/tasks/callback.py:
```diff
@@ -0,0 +1,206 @@
+class RunnerCallbackForProjectUpdate(RunnerCallback):
+    def __init__(self):
+        super(RunnerCallbackForProjectUpdate, self).__init__()
+        self.playbook_new_revision = None
```
Perfect, this looks great :+1:

AlanCoding commented on awx/main/tasks/callback.py:
```diff
@@ -0,0 +1,299 @@
+        # ansible-inventory and the awx.main.commands.inventory_import
+        # logger
+        if isinstance(self, RunnerCallbackForInventoryUpdate):
+            self.end_line = event_data['end_line']
```
It would be quite elegant to keep this line, lose the `if` and then defer to return the result from the super call.

amolgautam25 replied:
I have made the changes

AlanCoding commented on awx/main/tasks/callback.py:
```diff
@@ -0,0 +1,299 @@
+
+class RunnerCallbackForInventoryUpdate(RunnerCallback):
+    def __init__(self):
+        super(RunnerCallbackForInventoryUpdate, self).__init__()
```
shouldn't need `__init__` here.

amolgautam25 replied:
I have removed it.

AlanCoding commented on awx/main/tasks/callback.py:
```diff
@@ -0,0 +1,232 @@
+        # logger
+        self.end_line = event_data['end_line']
+
+        return super(RunnerCallbackForInventoryUpdate, self).event_handler(event_data)
```
This looks great! As an aesthetic comment, I would lose the long code comment because it's covered by the parent class.

amolgautam25 replied:
I have removed that, will be pushed in the next commit.

AlanCoding commented on awx/main/tasks/callback.py:
```diff
@@ -0,0 +1,232 @@
+                self.playbook_new_revision = returned_facts['scm_version']
+
+
+class RunnerCallbackForInventoryUpdate(RunnerCallback):
```
This looks perfectly structured to add the `event_data_key` here right below the class declaration

AlanCoding commented on awx/main/tasks/jobs.py:
```diff
@@ -622,21 +458,21 @@ def run(self, pk, **kwargs):
         extra_update_fields = {}
         fact_modification_times = {}
-        self.event_ct = 0
+        self.runner_callback.event_ct = 0
```
I think it would be best to move this to the `__init__` of the callback class, since it's just setting it to 0. That will also help with cases that we later reference `runner_callback.event_ct`, prevent attribute errors if this hasn't happened yet.

amolgautam25 replied:
Done , will be pushed in the next commit

AlanCoding commented on awx/main/tasks/jobs.py:
```diff
@@ -622,21 +458,21 @@ def run(self, pk, **kwargs):
         Needs to be an object property because status_handler uses it in a callback context
         '''
-        self.safe_env = {}
+        self.runner_callback.safe_env = {}
```
Again, I would say move this to the callback class `__init__`.

AlanCoding commented on awx/main/tasks/jobs.py:
```diff
@@ -669,15 +505,15 @@ def run(self, pk, **kwargs):
                     credential.credential_type.inject_credential(credential, env, self.safe_cred_env, args, private_data_dir)
 
-            self.safe_env.update(self.safe_cred_env)
+            self.runner_callback.safe_env.update(self.safe_cred_env)
```
You have a few things that need to be copied over from the task class to the callback class, but they're spread out throughout the `run` method. It would make the code easier to read if you consolidated them. Here would work. Those assignments look to be:
 - instance
 - parent_workflow_job_id
 - job_created
 - safe_env (assignment and then an update)

Just putting them all in 1 place would be nice. Maybe that will suggest that a helper method would help, but maybe not.

amolgautam25 replied:
I have made the changes , will be pushed in the next commit.

AlanCoding commented on awx/main/tasks/callback.py:
```diff
@@ -0,0 +1,213 @@
+    job_created = None
+    recent_event_timings = deque(maxlen=settings.MAX_WEBSOCKET_EVENT_RATE)
+    dispatcher = CallbackQueueDispatcher()
+    safe_env = {}
```
These are mutable data structures, and when you put them here (below the `class`), they are attached to the _class_. This is perfectly fine for immutable constants because they never change. This is a bad idea for anything that will be modified as you go. For those, it's better to set initial values in `__init__`. All of these look good, they just need to be saved to `self` in the init method so that they have a clearly defined scope of the lifetime of the RunnerCallback _object_, not the class. Classes are kind of immortal.

AlanCoding commented on awx/main/tasks/callback.py:
```diff
@@ -0,0 +1,247 @@
+    event_data_key = 'ad_hoc_command_id'
+
+    def __init__(self):
+        super(RunnerCallbackForJob, self).__init__()
```
You don't have the right class named passed here, it would be `RunnerCallbackForAdHocCommand` in this case, but it's better to delete these `__init__` methods that don't do anything other than call the super.

AlanCoding commented on awx/main/tasks/callback.py:
```diff
@@ -0,0 +1,247 @@
+    event_data_key = 'job_id'
+
+    def __init__(self):
+        super(RunnerCallbackForJob, self).__init__()
```
In this class, I would like to set `self.host_map` to `{}`. Same for the ad-hoc command class. This avoids throwing an AttributeError when referencing it before it has been assigned.

AlanCoding commented on awx/main/tasks/jobs.py:
```diff
@@ -444,171 +437,20 @@ def final_run_hook(self, instance, status, private_data_dir, fact_modification_t
+        if self.instance.spawned_by_workflow:
+            self.runner_callback.parent_workflow_job_id = self.instance.get_workflow_job().id
+
+        self.runner_callback.job_created = str(self.instance.created)
```
We need to avoid logic that might fail before the `try:` block starts. This looks good, all of these lines are good logic, I just don't want them _here_. It should work equally well around line 515.

That will help further consolidate all operations that add something to `self.runner_callback`. it's better to have them all in the same place as opposed to scattered about.

AlanCoding commented on awx/main/tasks/jobs.py:
```diff
@@ -1199,21 +1040,13 @@ class RunProjectUpdate(BaseTask):
-            returned_facts = returned_data.get('res', {}).get('ansible_facts', {})
-            if 'scm_version' in returned_facts:
-                self.playbook_new_revision = returned_facts['scm_version']
+        self.runner_callback = self.callback_class()
```
You shouldn't need this line, because it is done in `super(RunProjectUpdate, self).__init__(*args, **kwargs)`. That runs the `BaseTask` version of the init method.

When `BaseTask.__init__` runs, it will do this, `self.runner_callback = self.callback_class()`, and when it does that `self` is an instance of `RunProjectUpdate`, so it'll pick up the class property of `callback_class` so it should correctly make the project update version of the callback object.

AlanCoding commented on awx/main/tasks/callback.py:
```diff
@@ -0,0 +1,243 @@
+from awx.main.redact import UriCleaner
+from awx.main.constants import (
+    MINIMAL_EVENTS,
+)
```
syntax, put in single line.

AlanCoding commented on awx/main/tasks/callback.py:
```diff
@@ -0,0 +1,243 @@
+
+from awx.main.queue import CallbackQueueDispatcher
+
+logger = logging.getLogger('awx.main.tasks.jobs')
```
change logger jobs -> callback

AlanCoding commented on awx/main/tasks/callback.py:
```diff
@@ -0,0 +1,243 @@
+        if returned_data.get('task_action', '') == 'set_fact':
+            returned_facts = returned_data.get('res', {}).get('ansible_facts', {})
+            if 'scm_version' in returned_facts:
+                self.playbook_new_revision = returned_facts['scm_version']
```
this should return the same value as the super class, `r = super(...` and then `return r` at the end

AlanCoding commented on awx/main/tasks/callback.py:
```diff
@@ -0,0 +1,243 @@
+    event_data_key = 'inventory_update_id'
+
+    def __init__(self):
+        super(RunnerCallbackForInventoryUpdate, self).__init()
```
Here I would set `self.end_line` to 0

AlanCoding commented on awx/main/tasks/jobs.py:
```diff
@@ -2148,7 +1984,10 @@ class RunSystemJob(BaseTask):
+    callback_class = RunnerCallbackForSystemJob
+
+    def __init__(self):
+        self.runner_callback = self.callback_class()
```
Highest priority still TODO items seems like it would be to delete these `__init__` methods like this.

AlanCoding commented on awx/main/tasks/jobs.py:
```diff
@@ -800,7 +640,9 @@ class RunJob(BaseTask):
-    event_data_key = 'job_id'
+
+    def __init__(self):
+        super(RunJob, self).__init__()
```
shouldn't need this `__init__`, there's also another init method in the callback module that you can lose

AlanCoding commented on awx/main/tasks/callback.py:
```diff
@@ -0,0 +1,240 @@
+                self.instance.job_args = json.dumps(runner_config.command)
+                self.instance.job_cwd = runner_config.cwd
+                self.instance.job_env = job_env
+                self.instance.save()
```
for safety / performance, I do want to look into passing `update_fields` here. I think the updated fields are just these, right? job_args, job_cwd, job_env

amolgautam25 replied:
Yes , correct. 
I am looking at the original code and it seems only these 3 parameters are being changed.

AlanCoding commented on awx/main/tasks/callback.py:
```diff
@@ -0,0 +1,240 @@
+                from awx.main.signals import disable_activity_stream  # Circular import
+
+                with disable_activity_stream():
+                    self.instance = self.update_model(self.instance.pk, result_traceback=result_traceback)
```
whoops! we don't have `self.update_model`. You want to follow the same `self.instance.save()` pattern, but also probably include update fields, which would just be `result_traceback` here

fosterseth replied:
@AlanCoding 

do we need this `update_model` method at all, or can we scrap it? It's old (8 years +)

it has some retry logic in case of db errors, not sure how relevant that is anymore.

AlanCoding replied:
It's tied into how the unit tests work, and there is extra stuff about how the `modified` field is handled which we probably still need. I agree about the DB access. The philosophy is that the job prep and transmission should use the DB just like normal, but after receptor has received the job, the control process should maintain no active DB connections whatsoever. That's still a ways off, and I don't want to rush into refactoring the unit tests because I don't think we're ready for it.

AlanCoding commented on awx/main/tasks/callback.py:
```diff
@@ -0,0 +1,257 @@
+        '''
+        if event_data.get('event_data', {}).get('artifact_data', {}):
+            self.instance.artifacts = event_data['event_data']['artifact_data']
+            self.instance.save(update_fields=['artifacts'])
```
I feel like we should be consistent and if we're using update_model at all, we should use it here.

amolgautam25 replied:
Got it. I will go ahead and change it. However, this ```save``` method isn't put by me. It was there even before i started refactoring the code. So i thought to leave it as is.

jbradberry replied:
Red flag.  This should _not_ be making use of a function that is going to query a fresh copy out of the database.

AlanCoding commented on awx/main/tasks/utils.py:
```diff
@@ -0,0 +1,40 @@
+            time.sleep(5)
+            return update_model(model, pk, _attempt=_attempt + 1, **updates)
+        else:
+            logger.error('Failed to update %s after %d retries.', model._meta.object_name, _attempt)
```
@jbradberry I wanted to ask if you had any input on where a simple static method like this should go.

shanemcd replied:
It might be better to create a new file under https://github.com/ansible/awx/tree/devel/awx/main/utils

jbradberry replied:
I think I agree with Shane.

jbradberry commented on awx/main/tasks/utils.py:
```diff
@@ -0,0 +1,40 @@
+            return instance
+    except DatabaseError as e:
+        # Log out the error to the debug logger.
+        logger.debug('Database error updating %s, retrying in 5 ' 'seconds (retry #%d): %s', model._meta.object_name, _attempt + 1, e)
```
We have some scar tissue in this string from the Black conversion from a while back.  `5 ' 'seconds` -> `5 seconds`
